### ISSUE TITLE
Seemingly large memory consumption Keras models

### ISSUE THREAD

**Issue by ErikTromp** [deeplearning4j/deeplearning4j#3389]
#### Issue Description
**Observed behavior**
TL;DR - Suspiciously high memory usage for Keras models cause OoM

I am using DL4J to load pre-trained Keras models for VGG16 or (only 1 at a time) InceptionV3. I load the models into memory once and then use it to classify images, but I keep running out of memory all the time. It seems that just loading the models results in 4-5GB of RAM being used, Then once I start classifying images, it slowly rises to 7-8GB, causing my server to go out of memory.

**Expected behavior**
TL;DR - A lot less memory usage, no OoM
Not fully sure, it might be the high memory usage is normal? But since the models I am using are 90mb (InceptionV3) and 540mb (VGG16) on disk, I would hope to not use so much memory as I am seeing.

Links to gists showing my code are here:
https://gist.github.com/ErikTromp/d314f0c18c6f09070053fa69d8b9b6dc
https://gist.github.com/ErikTromp/9d8e674b40939018bfacee58398529d4
Files used for VGG16:
https://raw.githubusercontent.com/deeplearning4j/dl4j-examples/f9da30063c1636e1de515f2ac514e9a45c1b32cd/dl4j-examples/src/main/resources/trainedModels/VGG16.json
https://github.com/fchollet/deep-learning-models/releases/download/v0.1/vgg16_weights_th_dim_ordering_th_kernels.h5
Files used for InceptionV3
https://github.com/USCDataScience/dl4j-kerasimport-examples/blob/master/dl4j-import-example/data/inception-model.json
https://github.com/fchollet/deep-learning-models/releases/download/v0.2/inception_v3_weights_tf_dim_ordering_tf_kernels.h5

#### Version Information
Tried on Windows 10 and on Ubuntu 16.04.1 LTS
Using native (no GPU) DL4J binding, version 0.8.0

**Comment by saudet** [comment]
Could you check if this is used by the Java heap or not? We can get that value with `Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()`. Thanks!

**Comment by saudet** [comment]
BTW, Java tends to use as much memory as you let it, giving compute power in priority to the application over GC, but it will respect the higher limit returned by `Runtime.getRuntime().maxMemory()` that we can set with `java -Xmx` option.

**Comment by ErikTromp** [comment]
So I logged a bunch of those total mem - free mem just before loading the InceptionV3 model and right after. Here are some of the results I found, first number is *after* loading the model, second is *before*. Note: I completely restarted my application every time and did not do anything but boot it and then load InceptionV3.
```
948707408-591964952 = 340mb
1107843504-590086152 = 493mb
1692771672-586630648 = 1gb
2677843664-644751128 = ~2gb
2676544728-644723904 = ~2gb
949311392-592241824 = 340mb
```
Looking at this, apparently it isn't loading the model. I then suspected it would be using the model for classification, so I also printed the same numbers for before and after classification and I end up with using up 3-6mb per classification instance (which is probably due to getting the image's bytes and putting it on the model). Then after a while GC kicks in and frees up some stuff.

The result is that I am very puzzled now, I see a constant growth of memory consumption when I use the InceptionV3 model (or VGG16) for image classification, eventually yeilding OoM. My suspicions now are that data comes in too fast for it to handle since it's running on a not-so-super-fast linux server where classification of an image takes 0.5-1.5 seconds and data can _potentially_ come in faster than that.

I'll close this one for now and investigate further, thanks.

**Comment by saudet** [comment]
Are you getting a OOM in Java? If so, could you post the stack trace of that? It looks like something is using too much off-heap memory since the Java heap comes back down.

**Comment by ErikTromp** [comment]
Well the that's the thing, I am not getting any stack traces or logs anywhere. All that happens is that my JVM gets killed at some point. Since it takes a few minutes from startup to being killed I monitored it for a while and it rises to the max 7GB I gave the JVM (on an 8GB machine), swapping heavily just before getting killed, no trace or anything.

**Comment by ErikTromp** [comment]
So I tested it even further, on my own machine (32GB RAM, 50mb down/up), it seems to be stable at around 7-8GB with max 12GB allocated. That is my entire application (an analytics-intensive one, so yes those high memory usages are to be expected). It seems to stabilize back to 7-8GB after every GC once it hits 12GB RAM.

My conclusion is that the servers I ran it on have too little memory for my entire application, not just InceptionV3, but also are too slow either CPU- or Internetspeed-wise to keep up.

So I can only conclude again that it's nothing to do with DL4J or the model :)

**Comment by raver119** [comment]
How much memory you get allocated once only 1 model is loaded, and nothing else ever happened?

**Comment by ErikTromp** [comment]
Those were the figures I showed above, ranging from 340mb to 2gb - I reported earlier seeing 4-5GB of usage, but it seems quite random with a minimum of 340mb (obtained twice even), so I guess it's not at all that bad. (The h5 file is 90mb)

**Comment by raver119** [comment]
That's java side. Can you please check using ps axuw|grep PID_OF_YOUR_PROCESS ?

**Comment by saudet** [comment]
I see, try to set -Xmx4g instead. The same value is used for Java heap and
off-heap storage by default, so will total around 8g.

**Comment by tomthetrainer** [comment]
@saudet should this be flagged as an issue to document in our documentation as well as solving the problem from the original poster. 

Perhaps it is already documented but is it safe to say that loading and running inference on a NeuralNetwork will take the following amount of RAM, 

2* -Xmx value? 

So 4G leads to 8G, 2G leads to 4G..?

**Comment by huitseeker** [comment]
@tomthetrainer Yes, except there's gonna be complications when running on docker that should be mentioned in **BRIGHT RED CAPITAL LETTERS** the same place:
https://developers.redhat.com/blog/2017/03/14/java-inside-docker/
https://dzone.com/articles/java-and-memory-limits-in-containers-lxc-docker-an
http://trustmeiamadeveloper.com/2016/03/18/where-is-my-memory-java/

**Comment by saudet** [comment]
@tomthetrainer By default, in general, yes. It's mainly documented on the following two pages, not sure where would be a good general place to put thaw info though.
https://deeplearning4j.org/benchmark
https://deeplearning4j.org/spark

**Comment by ErikTromp** [comment]
Sorry I wasn't able to respond sooner. I ran the InceptionV3 model and again and got memory usage using ps. Booting my application, **before** loading InceptionV3 gives the following:
`root      8925 54.2  1.3 21967188 335136 pts/0 Sl   12:48   0:10 java -Duser.dir=/root/tuktu-1.3 -Dhttp.port=9008 -Xmx16G -cp  -jar /root/tuktu-1.3/lib/tuktu.tuktu-1.3-launcher.jar`
Then when I load InceptionV3 into memory, I get:
`root      8925  161  6.2 23361812 1535832 pts/0 Sl  12:48   2:56 java -Duser.dir=/root/tuktu-1.3 -Dhttp.port=9008 -Xmx16G -cp  -jar /root/tuktu-1.3/lib/tuktu.tuktu-1.3-launcher.jar`

**Comment by saudet** [comment]
I wonder if it isn't related to this somehow:
https://github.com/bytedeco/javacpp-presets/issues/423
Memory usage doesn't go down after calling System.gc() right?

**Comment by raver119** [comment]
cc @crockpotveggies ^^^^  O_o

**Comment by ErikTromp** [comment]
Not sure I follow - I keep the model in memory, so I am not surprised by seeing it use 1,5GB once loaded. I still think my OoM is coming from an overload of data or some other part of the app.

**Comment by saudet** [comment]
So, what's wrong with increasing the value of the -Xmx option then?

**Comment by ErikTromp** [comment]
Well the initial issue was that I run this on a node with just 8GB of RAM (not the one I posted the ps traces of), and that ones goes out of memory if I give it 7GB of RAM. But as said before, it's most likely due to the fact that the speed at which images come in is faster than it can classify them. I still think this issue can be closed.

**Comment by saudet** [comment]
Ah, if you're reading images with DataVec, then yes it might be very well caused by bytedeco/javacpp-presets#423

**Comment by ErikTromp** [comment]
Yes, I use the _org.datavec.image.loader.NativeImageLoader_ class. I looked into the other ticket, but couldn't find a direct solution there yet. Anything I can do about this now?

**Comment by saudet** [comment]
If that's the case, calling System.gc() in your loop somewhere might help
prevent allocating memory the native runtime won't release.

**Comment by ErikTromp** [comment]
@saudet I have had the time now to run it under various conditions with a GC inside the loop for every 25th image classification step. The outcome is that it now lasts quite a bit longer, but eventually still hits a ceiling. This even happens when I assign it 16GB RAM. I still don't get heap dumps or stack traces.

While this might be something in datavec, I still think it's because my application requires classification at a faster pace than the model can cope with. Would that sound reasonable?

**Comment by agibsonccc** [comment]
Have you tried with snapshots yet?

**Comment by saudet** [comment]
Ok, https://github.com/deeplearning4j/DataVec/pull/308 is probably going to do it, but it won't show up in the SNAPSHOT artifacts until we merge it.
